### PR TITLE
Feat/secure login

### LINKED ISSUE #2824: PIN/Biometric Login Screens
- User shall be able to enable PIN/Biometric Login when creating/restoring a wallet.
- User shall be to enable/disable PIN/Biometric Login from Settings Page.
- A lock screen shall be prompted to the user according to X time of app inactivity.
- User shall be set the time that is required for locking application during inactivity app.

### LINKED ISSUE #2830: Secure Login settings page

### PR DESCRIPTION
**Pull request by martinkaintas**
closes #2824
closes #2830

### CODE REVIEW

File: src/popup/App.vue [hunk 1]
```diff
@@ -229,6 +240,37 @@ export default defineComponent({
...
+    watch(
+      isAppActive,
+      async (isActive, wasActive) => {
+        if (!IS_MOBILE_APP) {
+          return;
+        }
+
+        // App went to background
+        if (wasActive && !isActive) {
+          lastTimeAppWasActive.value = new Date();
```
Comment (CedrikNikita) [inline src/popup/App.vue:256]: We can store `Date.now()`, and not call `getTime()` every time.

**Review by CedrikNikita — APPROVED** [review]

File: src/icons/lock.svg
Comment (peronczyk) [inline src/icons/lock.svg:1]: This icon is not pixel perfect which makes the icon "blurry". @onvisions can you update the icon?

File: src/popup/router/routes.ts [hunk 2]
```diff
@@ -442,6 +447,33 @@ export const routes: WalletAppRouteConfig[] = [
...
   },
+  {
+    path: '/more/settings/secure-login',
+    component: SecureLoginSettings,
+    name: ROUTE_SECURE_LOGIN_SETTINGS,
+    meta: {
+      title: 'secureLogin',
+      showHeaderNavigation: true,
+    },
+    beforeEnter: (to, from, next) => {
```
Comment (peronczyk) [inline src/popup/router/routes.ts:458]: I think user should be able to open the secure login settings even if used device does not support it.
I strongly suggest to use composables in the route definitions.

**Review by peronczyk — CHANGES_REQUESTED** [review]

File: src/popup/pages/Settings.vue [hunk 3]
```diff
@@ -46,10 +52,19 @@
...
+  useUi,
+} from '@/composables';
+import {
+  ROUTE_NETWORK_SETTINGS,
+  ROUTE_PERMISSIONS_SETTINGS,
+  ROUTE_SECURE_LOGIN_SETTINGS,
+} from '@/popup/router/routeNames';
 
 import PanelItem from '@/popup/components/PanelItem.vue';
+import { IS_MOBILE_APP } from '@/constants';
```
Comment (peronczyk) [inline src/popup/pages/Settings.vue:67]: Please move the constants import before composables import.

File: src/popup/pages/SecureLoginSettings.vue [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+    padding-inline: var(--screen-padding-x);
+    margin-bottom: 28px;
+  }
+
+  .text-description {
+    padding: 16px 0;
+  }
+
+  .options-wrapper {
+    border-top: 1px solid rgba(255, 255, 255, 0.15);
```
Comment (peronczyk) [inline src/popup/pages/SecureLoginSettings.vue:113]:
```suggestion
    border-top: 1px solid rgba($color-white, 0.15);
```

File: src/popup/pages/SecureLoginSettings.vue [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+  }
+
+  .options-info {
+    @include mixins.flex(flex-start, flex-start, column);
```
Comment (peronczyk) [inline src/popup/pages/SecureLoginSettings.vue:117]: I strongly suggest to avoid using this mixin as it gives no real value but makes the code less readable.

File: src/popup/pages/SecureLoginSettings.vue [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+      setSecureLoginEnabled,
+      setSecureLoginTimeout,
+      AUTHENTICATION_TIMEOUTS,
+    };
+  },
+});
+</script>
+
+<style lang="scss" scoped>
+@use '../../styles/variables';
```
Comment (peronczyk) [inline src/popup/pages/SecureLoginSettings.vue:98]:
```suggestion
@use '../../styles/variables' as *;
```

File: src/popup/pages/SecureLoginSettings.vue (hunk 4 above)
Comment (peronczyk) [reply src/popup/pages/SecureLoginSettings.vue:113]: Also I sugest to take a look on global styles applied to `hr` element (global.scss). This line is used in `NetworkForm.vue`.

File: src/popup/pages/SecureLoginSettings.vue [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+      <div class="secure-login-settings">
+        <div class="top-wrapper">
+          <p class="text-description">
+            {{ $t('pages.secureLogin.description') }}
+          </p>
+
+          <SwitchButton
+            :label="$t('pages.secureLogin.enableSecureLogin')"
+            :model-value="isSecureLoginEnabled"
+            @update:modelValue="setSecureLoginEnabled"
```
Comment (peronczyk) [inline src/popup/pages/SecureLoginSettings.vue:13]: Maybe it would be possible to use:
```suggestion
            v-model="isSecureLoginEnabled"
```

File: src/popup/components/Modals/SecureLogin.vue [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+      margin-bottom: 4px;
+    }
+
+    .text {
+      @extend %face-sans-16-regular;
+      line-height: 24px;
+      color: rgba($color-white, 0.85);
+    }
+  }
+
```
Comment (peronczyk) [inline src/popup/components/Modals/SecureLogin.vue:125]: This line can be removed :)

File: src/popup/components/Modals/SecureLogin.vue [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+    from-bottom
+    full-screen
+    no-padding
+    centered
+    transparent
+    @close="handleClose"
+    @click="initAuthenticate"
+  >
+    <div
+      v-if="!isAuthenticated"
```
Comment (peronczyk) [inline src/popup/components/Modals/SecureLogin.vue:13]: Consider displaying something in case user is Authenticated.

File: src/popup/components/Modals/EnableSecureLogin.vue [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+    margin-top: 44px;
+    opacity: 0.5;
+  }
+
+  .info {
+    .title {
+      color: $color-white;
+      padding:16px 0;
+
+      @extend %face-sans-18-medium;
```
Comment (peronczyk) [inline src/popup/components/Modals/EnableSecureLogin.vue:120]: Please put the extends at the beginning of style definitions.

File: src/popup/components/Modals/EnableSecureLogin.vue [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+    }
+
+    .text {
+      color: rgba($color-white, 0.85);
+      line-height: 20px;
+      gap: 10px;
+
+      @extend %face-sans-14-light;
+
+      @include mixins.flex(center, center, column);
```
Comment (peronczyk) [inline src/popup/components/Modals/EnableSecureLogin.vue:130]: I strongly advise against using this mixin as it gives more harm than real value.

File: src/popup/components/Modals/EnableSecureLogin.vue [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+
+      @extend %face-sans-18-medium;
+    }
+
+    .text {
+      color: rgba($color-white, 0.85);
+      line-height: 20px;
+      gap: 10px;
+
+      @extend %face-sans-14-light;
```
Comment (peronczyk) [inline src/popup/components/Modals/EnableSecureLogin.vue:128]: Please put this at the beginning of style definitions.

File: src/popup/components/Modals/EnableSecureLogin.vue [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+
+  .icon-wrapper {
+    margin-top: 44px;
+    opacity: 0.5;
+  }
+
+  .info {
+    .title {
+      color: $color-white;
+      padding:16px 0;
```
Comment (peronczyk) [inline src/popup/components/Modals/EnableSecureLogin.vue:118]:
```suggestion
      padding: 16px 0;
```

File: src/popup/components/Modal.vue [hunk 14]
```diff
@@ -258,6 +260,16 @@ export default defineComponent({
     }
   }
 
+  &.transparent {
+    backdrop-filter: blur(10px);
```
Comment (peronczyk) [inline src/popup/components/Modal.vue:264]: This rule is already applied with `blur-bg` class.

File: src/popup/components/Modal.vue [hunk 15]
```diff
@@ -258,6 +260,16 @@ export default defineComponent({
...
+    background-color: rgba(variables.$color-black, 0.8);
```
Comment (peronczyk) [inline src/popup/components/Modal.vue:265]: Is this really necessary as the default value is the same but the opacity is `0.7` instead of `0.8`?

**Comment by Liubov-crypto** [comment]
LGTM in general. Leaved 1 question to @onvisions

File: src/popup/components/IconBoxed.vue [hunk 16]
```diff
@@ -12,9 +16,23 @@
...
+] as const;
+
+export type IconBoxedVariant = typeof ICON_BOXED_VARIANT[number];
+
 export default defineComponent({
   props: {
     icon: { type: Object, default: null },
+    smallIcon: Boolean,
+    variant: {
+      type: String,
```
Comment (peronczyk) [inline src/popup/components/IconBoxed.vue:32]:
```suggestion
      type: String as PropType<IconBoxedVariant>,
```

File: src/popup/components/IconBoxed.vue [hunk 17]
```diff
@@ -1,9 +1,13 @@
 <template>
-  <div class="icon-boxed">
+  <div
+    class="icon-boxed"
+    :class="{ [variant]: variant }"
```
Comment (peronczyk) [inline src/popup/components/IconBoxed.vue:4]: Consider this change:
```suggestion
    :class="[ variant ]"
```

**Review by peronczyk — CHANGES_REQUESTED** [review]

File: src/composables/biometricAuth.ts [hunk 18]
```diff
@@ -0,0 +1,83 @@
...
+      openModal(MODAL_ENABLE_SECURE_LOGIN);
+    }
+  }
+
+  (async () => {
+    if (composableInitialized) {
+      return;
+    }
+    await checkIsAvailable();
+    composableInitialized = true;
```
Comment (peronczyk) [inline src/composables/biometricAuth.ts:71]: Setting this after awaiting is in my opinion wrong, because multiple components could run this composable in parallel and all `checkIsAvailable` method would be fired multiple times.

File: src/composables/biometricAuth.ts [hunk 19]
```diff
@@ -0,0 +1,83 @@
...
+  /**
+   * Prompts the user to authenticate using biometric authentication.
+   * Returns a promise that resolves when the user is authenticated
+   * or if biometric authentication is not available.
+   */
+  async function authenticate(): Promise<void> {
+    return new Promise((resolve, reject) => {
+      if (!isAvailable.value || !isSecureLoginEnabled.value) {
+        resolve();
+        return;
```
Comment (peronczyk) [inline src/composables/biometricAuth.ts:30]:
```suggestion
        return resolve();
```

File: src/composables/ui.ts [hunk 20]
```diff
@@ -31,6 +36,10 @@ const hiddenCards = useStorageRef<string[]>(
     ],
   },
 );
+const secureLogin = useStorageRef<ISecureLoginSettings>(
+  {},
+  STORAGE_KEYS.secureLogin,
+);
```
Comment (peronczyk) [inline src/composables/ui.ts:42]: In order to decrease the amount of future migrations and amount of separate storage entries consider moving the secure login settings to other settings.
I know that "other" is not the best name for these settings. Maybe in the future we could use something like `primarySettings` or `commonSettings`.

File: src/popup/App.vue [hunk 21]
```diff
@@ -134,9 +138,16 @@ export default defineComponent({
...
+    const {
+      checkIsAvailable,
+      isAuthenticated,
+      openAuthModal,
+      deauthenticate,
+    } = useBiometricAuth();
 
     const innerElement = ref<HTMLDivElement>();
     const delayedShowHeader = ref(false);
+    const lastTimeAppWasActive = ref<Date>();
```
Comment (peronczyk) [inline src/popup/App.vue:150]: I suggest to move this to `ui` composable.

File: src/popup/App.vue [hunk 22]
```diff
@@ -229,6 +240,37 @@ export default defineComponent({
...
+        if (isActive && !wasActive) {
+          if (isSecureLoginEnabled.value && isAuthenticated.value && lastTimeAppWasActive.value) {
+            await checkIsAvailable(); // Check if biometric auth is still available
+            const now = new Date();
+            const elapsedTime = now.getTime() - lastTimeAppWasActive.value.getTime();
+            if (elapsedTime > secureLoginTimeout.value) {
+              deauthenticate();
+              await openAuthModal();
+            }
+          }
```
Comment (peronczyk) [inline src/popup/App.vue:269]: In my opinion this logic should be moved to `useBiometricAuth` composable.

File: src/composables/biometricAuth.ts [hunk 23]
```diff
@@ -0,0 +1,83 @@
...
+import { tg as t } from '@/popup/plugins/i18n';
+import { IS_MOBILE_APP, MODAL_ENABLE_SECURE_LOGIN, MODAL_SECURE_LOGIN } from '@/constants';
+import { useUi } from './ui';
+import { useModals } from './modals';
+
+let composableInitialized = false;
+const isAvailable = ref(false);
+const isAuthenticated = ref(false);
+
+export function useBiometricAuth() {
```
Comment (peronczyk) [inline src/composables/biometricAuth.ts:13]: What do you think about making it more generic by not focusing only on the biometric aspect? This way we could extend this composable with different methods of athenticating users.
```suggestion
export function useAuth() {
```

File: src/popup/router/index.ts [hunk 24]
```diff
@@ -76,6 +75,10 @@ router.beforeEach(async (to, from, next) => {
     return;
   }
 
+  if (isSecureLoginEnabled.value && !isAuthenticated.value) {
+    await openAuthModal();
+  }
```
Comment (peronczyk) [inline src/popup/router/index.ts:80]: As this test is repeated few times in the PR consider creating some kind of method that would do the check and eventually open the modal: `checkUserAuth()`

**Review by peronczyk — CHANGES_REQUESTED** [review]

File: src/popup/router/routes.ts (hunk 2 above)
Comment (martinkaintas) [reply src/popup/router/routes.ts:458]: @onvisions What do you think?

File: src/popup/pages/SecureLoginSettings.vue (hunk 7 above)
Comment (martinkaintas) [reply src/popup/pages/SecureLoginSettings.vue:13]: Since it is a computed value we cannot

File: src/popup/components/Modals/SecureLogin.vue (hunk 9 above)
Comment (martinkaintas) [reply src/popup/components/Modals/SecureLogin.vue:13]: If user is Authenticated this modal should disappear

File: src/popup/components/Modal.vue (hunk 15 above)
Comment (martinkaintas) [reply src/popup/components/Modal.vue:265]: According to the design it's 0.8🤷‍♂️ 
@onvisions What do you think?

File: src/icons/lock.svg
Comment (onvisions) [reply src/icons/lock.svg:1]: Updated resources were provided. However blurring issue might be related to a visual defect causing all modals to look blurred within Chrome browser extension. I mentioned this several times and I consider it important to investigate and fix. Not for this release though.

File: src/popup/components/Modal.vue (hunk 15 above)
Comment (onvisions) [reply src/popup/components/Modal.vue:265]: We can leave the default value then. Let's make it 0.7.

File: src/popup/router/routes.ts (hunk 2 above)
Comment (onvisions) [reply src/popup/router/routes.ts:458]: I think the user should not even see "Secure login" tab in Settings of the _browser extension_ because currently we do not support secure login for the extension, we do only for the apps. 
It will be a different story if we introduce the feature with password or PIN for the browser extension too. 

But currently it doesn't make sense to me: As a user I see the option "Secure login" in my browser extension Settings and I click because I'd like to have it enabled. I go to Secure login page and everything there is disabled. So now I am starting to wonder why... Is there a problem with my browser, OS. It's a waste of time and creates consfusion.

Another approach will be to display message that Secure login is currently supported only for the wallet mobile apps and not for the web version or browser extensions. However I think all this is an unnecessary complication of the feature and we'd better think of adding some secure authentication support for the web and extension users. @smaroudasunicorn

File: src/composables/auth.ts [hunk 25]
```diff
@@ -0,0 +1,117 @@
...
+        }).then(() => {
+          isAuthenticated.value = true;
+        }));
+      } catch (error) {
+        reject(error);
+      }
+    });
+  }
+
+  function deauthenticate() {
```
Comment (peronczyk) [inline src/composables/auth.ts:55]: Although there is a `deauthenticate` word, but it relates one specific thing: Wi-Fi. I suggest to use something different.

After doing some research I found out that most common actions on the `authenticated` state is the `login` / `logout`, `signIn` / `signOut`. For example take a look here: https://auth0.com/docs/quickstart/spa/vanillajs/01-login#evaluate-the-authentication-state

File: src/composables/auth.ts [hunk 26]
```diff
@@ -0,0 +1,117 @@
...
+              await openAuthModal();
+            }
+          }
+        },
+      );
+    });
+  })();
+
+  return {
+    isAvailable: readonly(isAvailable),
```
Comment (peronczyk) [inline src/composables/auth.ts:109]: I think this should not be available from the outside as we are not always sure if we already set the value.

File: src/composables/auth.ts [hunk 27]
```diff
@@ -0,0 +1,117 @@
...
+const isAvailable = ref(false);
+const isAuthenticated = ref(false);
+
+export function useAuth() {
+  const { isSecureLoginEnabled } = useUi();
+  const { openModal } = useModals();
+
+  async function checkIsAvailable() {
+    isAvailable.value = (await BiometricAuth.checkBiometry()).isAvailable;
+  }
```
Comment (peronczyk) [inline src/composables/auth.ts:24]: I suggest to use similar approach as `getAeSdk` so we set the value only on first call and then return it.

File: src/composables/auth.ts [hunk 28]
```diff
@@ -0,0 +1,117 @@
...
+  effectScope,
+} from 'vue';
+
+import { tg as t } from '@/popup/plugins/i18n';
+import { IS_MOBILE_APP, MODAL_ENABLE_SECURE_LOGIN, MODAL_SECURE_LOGIN } from '@/constants';
+import { useUi } from './ui';
+import { useModals } from './modals';
+
+let composableInitialized = false;
+const isAvailable = ref(false);
```
Comment (peronczyk) [inline src/composables/auth.ts:15]: I think this could be a regular `let` variable instead of reactive reference.

File: src/composables/auth.ts [hunk 29]
```diff
@@ -0,0 +1,117 @@
...
+    isAvailable.value = (await BiometricAuth.checkBiometry()).isAvailable;
+  }
+
+  /**
+   * Prompts the user to authenticate using biometric authentication.
+   * Returns a promise that resolves when the user is authenticated
+   * or if biometric authentication is not available.
+   */
+  async function authenticate(): Promise<void> {
+    return new Promise((resolve, reject) => {
```
Comment (peronczyk) [inline src/composables/auth.ts:32]: I don't see big reason for wrapping the logic with additional `Promise`.

**Review by peronczyk — CHANGES_REQUESTED** [review]

File: src/popup/pages/Settings.vue [hunk 30]
```diff
@@ -6,6 +6,12 @@
           :to="{ name: 'settings-seed-phrase' }"
           :title="$t('pages.index.seedPhrase')"
         />
+        <PanelItem
+          v-if="IS_MOBILE_APP"
+          :to="{ name: ROUTE_SECURE_LOGIN_SETTINGS }"
+          :info="isSecureLoginEnabled ? 'On' : 'Off'"
```
Comment (peronczyk) [inline src/popup/pages/Settings.vue:12]: Maybe those two vallues could be taken from translations?

File: src/composables/auth.ts [hunk 31]
```diff
@@ -0,0 +1,117 @@
...
+        reject(error);
+      }
+    });
+  }
+
+  function deauthenticate() {
+    isAuthenticated.value = false;
+  }
+
+  async function openAuthModal() {
```
Comment (peronczyk) [inline src/composables/auth.ts:59]: In my opinion the function name should reflect the modal name. So if modal name is "Foo" the method should be "OpenFooModal". So if I see the method name somewhere I can immediately know which modal file I can open to dig into the code.

File: src/composables/auth.ts [hunk 32]
```diff
@@ -0,0 +1,117 @@
...
+    await checkIsAvailable(); // Check if biometric auth is still available
+    if (isAvailable.value && isSecureLoginEnabled.value && !isAuthenticated.value) {
+      await openModal(MODAL_SECURE_LOGIN);
+    }
+  }
+
+  async function openEnableAuthModal() {
```
Comment (peronczyk) [inline src/composables/auth.ts:66]: IDEM

File: src/composables/auth.ts [hunk 33]
```diff
@@ -0,0 +1,117 @@
...
+
+  (async () => {
+    if (composableInitialized) {
+      return;
+    }
+    composableInitialized = true;
+    await checkIsAvailable();
+
+    // Create persistent effect scope to avoid watcher being disposed
+    effectScope(true).run(() => {
```
Comment (peronczyk) [inline src/composables/auth.ts:80]: I suggest to use `createCustomScopedComposable` instead.

**Review by peronczyk — CHANGES_REQUESTED** [review]

File: src/popup/router/routes.ts (hunk 2 above)
Comment (peronczyk) [reply src/popup/router/routes.ts:458]: I think I did wrong explanation so let me rephrase.

What I was suggesting is to not write any router level code that would block user from manually entering the Secure Login settings page as it makes sense because we are removing the button that allows tthe access to this page.

If user enters the page address of this settings page in the address bar he should access the page with an information that this feature is not available to him. Right now the `SecureLoginSettings` is not testing if user is actually able to use this feature which is bad in my opinion.

Also I forgot to add **no** to this sentence:
> I strongly suggest to use composables in the route definitions.

So it should be: I strongly suggest to **NOT** use composables in the route definitions.

File: src/popup/router/routes.ts [hunk 34]
```diff
@@ -4,7 +4,10 @@ import {
   PROTOCOL_VIEW_TRANSACTION_DETAILS,
   PROTOCOL_VIEW_ACCOUNT_DETAILS_ASSETS,
   PROTOCOL_VIEW_ACCOUNT_DETAILS_NAMES,
+  UNFINISHED_FEATURES,
+  IS_MOBILE_APP,
 } from '@/constants';
+import { useAuth } from '@/composables';
```
Comment (peronczyk) [inline src/popup/router/routes.ts:10]: I strongly suggest to NOT use composables in the route definitions file :)

File: src/icons/lock.svg
Comment (peronczyk) [inline src/icons/lock.svg:1]: I think the fill color should be taken from our CSS color variables. This will allow us to change color themes in the future.

File: src/popup/components/IconBoxed.vue [hunk 35]
```diff
@@ -1,5 +1,8 @@
 <template>
-  <div class="icon-boxed">
+  <div
+    class="icon-boxed"
+    :class="[variant]"
```
Comment (peronczyk) [inline src/popup/components/IconBoxed.vue:4]: To make it easier to debug I suggest to:
```suggestion
    :class="[`variant-${variant}`]"
```

File: src/composables/auth.ts [hunk 36]
```diff
@@ -0,0 +1,120 @@
...
+   * Prompts the user to authenticate using biometric authentication.
+   * Returns a promise that resolves when the user is authenticated
+   * or if biometric authentication is not available.
+   */
+  async function authenticate(): Promise<void> {
+    if (!isAvailable || !isSecureLoginEnabled.value) {
+      return Promise.resolve();
+    }
+    try {
+      return (BiometricAuth.authenticate({
```
Comment (peronczyk) [inline src/composables/auth.ts:52]: I think the rounded brackets around returned value can be emoved.

File: src/popup/router/routes.ts (hunk 2 above)
Comment (onvisions) [reply src/popup/router/routes.ts]: In this case I'll add the Secure login screen mockup for the case when the page is not accessible from Settings but still accessible directly via URL.I'm syncing it with Martin for the implementation.

File: src/popup/components/Modals/SecureLogin.vue (hunk 9 above)
Comment (peronczyk) [reply src/popup/components/Modals/SecureLogin.vue:13]: OK, so why adding `v-if` for this case if other cases won't happen?

**Review by peronczyk — CHANGES_REQUESTED** [review]

File: src/popup/router/routes.ts (hunk 2 above)
Comment (peronczyk) [reply src/popup/router/routes.ts]: I wouldn't do anything complex. We have our `InfoBox` component so we need only some kind of text to be put there.

**Review by peronczyk — APPROVED** [review]

**Comment by Liubov-crypto** [comment]
LGTM
